content/blog/2025-10-02-50-million-zsets/index.md

```diff
@@ -0,0 +1,65 @@
+
+One such workload is [Raider.IO](https://raider.io), the most popular companion site for World of Warcraft players. Raider.IO maintains leaderboards that track millions of characters and guilds, updating instantly whenever a player clears a dungeon or raid. That translates into hundreds of millions of sorted set operations every day. In this environment, downtime or latency isn't acceptable. Memory efficiency and throughput directly shape the player experience. Running Valkey at hyperscale for workloads like Raider.IO means that improvements in Valkey 8.1 immediately impact both infrastructure cost and the ability to absorb peak traffic without breaking a sweat.
+
+## Benchmark Setup
+
+To provide a fair comparison, both Valkey 8.0 and Valkey 8.1.1 were run on the same hardware:
+
+* AWS c8g.2xl instance (Graviton4, 8 vCPU, 16 GB RAM)
+* Identical settings for persistence, I/O threads and memory limits
+* 50 million members inserted into a ZSET via pipelined ZADD commands (score=i, member=m:{i})
+* 250,000 item batch size per pipeline
+* Metrics collected every 1,000,000 inserts (flush between runs)
+
+The benchmark code is open sourced and straightforward to reproduce: it connects to both servers, flushes the test key, performs batched inserts, and records `used_memory`, `used_memory_rss`, total elapsed time, and throughput after each million inserts. This repeatability mirrors the ethos of Valkey's community - every optimization is measurable, and anyone can verify the results.
+
+## Memory Usage – 8.1 vs 8.0
+
+Valkey 8.1's redesigned dictionary structure cuts [roughly 20–30 bytes per key](/blog/valkey-8-1-0-ga), and those savings compound with scale. The figure below charts `used_memory` during the benchmark for both versions:
+
+![Chart illustrating used vs inserts. It shows Valkey 8.1 with lower used memory](used-memory-vs-inserts.png)
+
+At 1 million inserts, Valkey 8.0 used ~95 MB while Valkey 8.1 used ~81 MB. As the ZSET grew, the gap widened. By 10 million inserts, 8.0 consumed 1.06 GB versus 0.77 GB for 8.1 - **a 27% reduction**. At the end of the run (50 million inserts), 8.1 used 3.77 GB compared to 4.83 GB on 8.0, saving 1.06 GB (≈22 %).
+
+These numbers align with the release notes. Valkey's 8.1 announcement highlights lower per‑key overheads and improved data structure handling; Linuxiac notes that 8.1's architectural changes can reduce memory footprints by [approximately 20 bytes per KV pair](https://linuxiac.com/valkey-8-1-in-memory-data-store-unleashes-10-faster-throughput), with each pair normally consuming 100 bytes (a 20% reduction!). Our results confirm those claims on a large ZSET workload.
```

Comment:
```suggestion
These numbers align with the release notes. Valkey's 8.1 announcement highlights lower per‑key overheads and improved data structure handling; Linuxiac notes that 8.1's architectural changes can reduce memory footprints by [approximately 20 bytes per KV pair](https://linuxiac.com/valkey-8-1-in-memory-data-store-unleashes-10-faster-throughput), with each pair normally consuming 100 bytes (a 20% reduction!). Our results confirm those claims on a large ZSET workload.
```

These are apples to oranges comparisons, because the Linuxiac is talking per key and you are talking about per item overhead. I think that's fine, I think it's more like "Our results also confirm a similar improvement on large ZSET workloads. "